zerver/lib/converters.py

```diff
@@ -0,0 +1,269 @@
+
+    if isinstance(exc, KeyError):
+        return _("Missing required field")
+    elif isinstance(exc, TypeError) and exc.args[0].endswith("object is not iterable"):
+        # There are in fact other types like set, deque, and tuple that can
+        # raise this error. We need cattrs to extend error handling so that
+        # we have the target type to distinguish these cases. For now,
+        # because our codebase don't demand these types as the target, it is
+        # fine to emit this error message.
+        return _("Expected a list")
+    elif isinstance(exc, (ValueError, TypeError)):
+        if target_type is int:
+            return _("Expected an integer")
+        elif target_type is str:
+            return _("Expected a string")
+        elif target_type is bool:
+            return _("Expected a boolean")
+        elif target_type is float:
+            return _("Expected a float")
+    elif isinstance(exc, ForbiddenExtraKeysError):
+        return _("Unexpected arguments: {keys}").format(keys=", ".join(exc.extra_fields))
+    elif isinstance(exc, AttributeError) and exc.args[0].endswith(
+        "object has no attribute 'items'"
+    ):
+        # This was supposed to be a mapping (and have .items())
+        return _("Expected a dict")
+    elif isinstance(exc, AttributeError) and exc.args[0].endswith("object has no attribute 'copy'"):
+        # This was supposed to be a mapping (and have .copy())
+        # Used for TypedDicts.
+        return _("Expected a dict")
+    elif "not in literal" in exc.args[0]:
+        # This applies to errors related to structuring Literals.
+        # TODO: Get a PR for cattrs to avoid inspecting the error message.
+        value = exc.args[0].split("not in literal")[0].strip()
+        return _("Invalid value: {value}").format(value=value)
+    elif isinstance(exc, ClientFacingError):
+        return exc.args[0]
+
+    # Unhandled errors are unexpected so they must be re-raised.
+    raise exc
+
+
+def structure_isinstance(value: object, typ: Type[T]) -> T:
+    if is_optional(typ):
+        type_args = get_args(typ)
+        typ = type_args[0] if type_args[1] is type(None) else type_args[1]
+        if value is None:
+            return value  # type: ignore[return-value] # If typ is optional, value being None is accepted
+    if not isinstance(value, typ):
+        exc = TypeError(f"{value} is not a {typ.__name__}")
+        # Associate a note with the TypeError to give it a bit more context and
+        # type information, so that we can produce better error messages in the
+        # format_exception helper.
+        note = ZulipValidationNote("primitive value failed isinstance check", typ)
+        exc.__notes__ = [*getattr(exc, "__notes__", []), note]  # type: ignore[attr-defined] # This is a PEP-678 feature will be added in Python 3.11. TODO: use exc.add_note once it is supported.
+        raise exc
+    return value
+
+
+def structure_general(value: object, typ: Type[T]) -> T:
```

Comment:
I don't have huge objections to the use of "typ" as a variable, since it does have the advantage of brevity and it's pretty easy to understand from context.  But I would consider using things like "generic_type" or "primitive_type" depending on the context.

Reply:
Yeah, `target_type` would be better.